src/Lean/Parser/Basic.lean

```diff
@@ -660,6 +660,9 @@ def quotedCharCoreFn (isQuotable : Char → Bool) (inString : Bool) : ParserFn :
       andthenFn hexDigitFn hexDigitFn c (s.next' input i h)
     else if curr == 'u' then
       andthenFn hexDigitFn (andthenFn hexDigitFn (andthenFn hexDigitFn hexDigitFn)) c (s.next' input i h)
+    else if curr == 'U' then
+      (andthenFn hexDigitFn <| andthenFn hexDigitFn <| andthenFn hexDigitFn <| andthenFn hexDigitFn <|
+        andthenFn hexDigitFn <| andthenFn hexDigitFn <| andthenFn hexDigitFn hexDigitFn) c (s.next' input i h)
```

Comment:
One thing missing is a check that this is less than 0x110000 and not in the range [0xd800, 0xdfff]. That said, there's a missing check for `u` that it's not in the surrogate range [0xd800, 0xdfff]. Right now, `'\ud800'` gives `'\x00'`.

Reply:
> Right now, `'\ud800'` gives `'\x00'`

I guess this is related to #5445. Is `Char.ofNat 0xd80f = 0` really desirable? Python seems to make a different choice here.

Reply:
Python passes these pairs through, right? And `"\U00110000"` is a syntax error?